### PR TITLE
PAPER-3: Photonic version of QTRL

### PR DESCRIPTION
**Pull request by leithkarrai**
This code is the photonic version of this paper : https://arxiv.org/html/2407.06103v1 
using MerLin.

### CODE REVIEW

File: papers/QTRL/requirements.txt [hunk 1]
```diff
@@ -0,0 +1,5 @@
+merlinquantum==0.2.3
+torch
+numpy
```
Comment (CassNot) [inline papers/QTRL/requirements.txt]: torch and numpy not necessary as installed upon installation of merlinquantum

File: papers/QTRL/configs/defaults.json [hunk 2]
```diff
@@ -0,0 +1,11 @@
+{
```
Comment (CassNot) [inline papers/QTRL/configs/defaults.json:1]: the `final_output_size` should be hardcoded with respect to the `env_name` as it directly depends on it
Here, if I change the `final_output_size`, the code will fail
It can be fixed in the `runner.py`

File: papers/QTRL/lib/util.py [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+        
+
+        torch.nn.utils.clip_grad_norm_(hybrid_model.parameters(), max_norm=1.0)
+        
+        optimizer.step()
+        
+
+        total_reward = sum(rewards)
+        print(f"Episode {episode:4d} | rewards : {total_reward:5.1f} | Loss : {policy_loss.item():.4f}")
+
```
Comment (CassNot) [inline papers/QTRL/lib/util.py:288]: It could be nice to add the max reward that makes sense for the specific environment, to see if the training worked !

File: papers/QTRL/lib/util.py [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+            low=0.0, high=1.0, shape=(self.flat_dim,), dtype=np.float32
+        )
+
+    def observation(self, obs):
+        image = obs['image']
+        flat_image = image.flatten().astype(np.float32) / 10.0
+        return flat_image
+
+
+def train_minigrid_qtrl(hybrid_model, num_episodes=1000, learning_rate=0.001, seed=42):
```
Comment (CassNot) [inline papers/QTRL/lib/util.py]: Is there a way to not duplicate the training loop for both environment ?

File: papers/QTRL/README.md [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+```text
+papers/QTRL/
+├── cli.json                # CLI argument definitions and mapping
+├── configs/
+│   └── defaults.json       # Default hyperparameters
+├── lib/
+│   ├── runner.py           # Main entry point (train_and_evaluate)
+│   ├── util.py             # Quantum circuits and RL training logic
+│   ├── CartPole_photonic.py # Specific logic for CartPole
```
Comment (CassNot) [inline papers/QTRL/README.md]: these files do not exist (CarPole and Minigrid)

**Review by CassNot — CHANGES_REQUESTED** [review]
Good work ! I added some comments and here are some other guidelines/comments: 
- torchquantum backend: could you add the original torchquantum layers ? It could be an additional parameter in the config file `"backend"` to be either `"merlin"` or `"torchquantum"`. This way, we could easily run the different options
- tests: you should add tests in a `tests/ `folder to test your code (unit tests for instance). This would ensure sanity checking of the repo when MerLin or other packages undergo updates. You should then add `pytest` in the requirements

**Comment by leithkarrai** [comment]
@claude review

File: papers/QTRL/lib/util.py [hunk 6]
```diff
@@ -0,0 +1,527 @@
...
+
+import gymnasium as gym
+import numpy as np
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+import torch.optim as optim
+import torchquantum.functional as tqf
+from gymnasium.core import ObservationWrapper
+from lib.torchmps import MPS
```
Comment (CassNot) [inline papers/QTRL/lib/util.py:12]: **High**: papers/QTRL/lib/torchmps/ is ignored in the gitignore and therefore, torchmps cannout be found. 

In the DQNN paper, the torchmps is only a few files and is not heavy and tests were removed. You could handle it as in DQNN

File: papers/QTRL/configs/defaults.json (hunk 2 above)
Comment (CassNot) [reply papers/QTRL/configs/defaults.json:1]: This is now addressed in `runner.py` as it computes `state_dim * action_dim` and passes that into model creation. 
However, defaults.json (as the cli.json and README.md but I am grouping these 3 cases in this comment) still expose `final_output_size`. 
As it is now ignored, changing it silently does nothing. 
You should remove it from config/CLI/docs

File: papers/QTRL/configs/defaults.json [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
```
Comment (CassNot) [inline papers/QTRL/configs/defaults.json:1]: **Medium**: classic is given as a backend in the README, but `util.py` requires `args["layer_dim"]`, while defaults do not define it. `python implementation.py --paper QTRL --backend classic` therefore hit a `KeyError` unless the user knows to pass `--layer_dim`
Maybe a way to address this problem could be to either add a valid `layer_dim` to `defaults.json`, or make classic fail with a clear validation error before model construction. 
As `classic` is a supported CLI choice, I’d prefer the option of adding it explicitly to `defaults.json` and documenting it

**Review by CassNot — CHANGES_REQUESTED** [review]
Good updates but some are still needed: 
- **High**:`torchquantum` folder is not needed here: it is very heavy and you can add the torchquantum in the requirements as a git install line: `-e git+https://github.com/mit-han-lab/torchquantum.git#egg=torchquantum` (as in the qllm for instance)
- the other are given as comments in the related files

File: papers/QTRL/lib/util.py [hunk 8]
```diff
@@ -0,0 +1,527 @@
...
+import random
+
+import gymnasium as gym
+import numpy as np
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+import torch.optim as optim
+import torchquantum.functional as tqf
+from gymnasium.core import ObservationWrapper
```
Comment (CassNot) [inline papers/QTRL/lib/util.py:11]: Since `torchquantum` is only meant to back one optional backend (`--backend torchquantum`), but the import is at module scope, running the advertised default command (`python implementation.py --paper QTRL`, backend merlin_mlp) will raise `ImportError` for anyone who hasn't separately installed `TorchQuantum`
I recommend a lazy/local import inside TorchQuantumModel so it will be imported only if needed

**Review by CassNot — CHANGES_REQUESTED** [review]
This implementation fails the repo's **mandatory structure** (how_to_contribute.md) (for example, you do not have **tests**)
In the required structure, there is also a **notebook**
Your README needs to explain how your model performs compared to the original one

We need tests to ensure the sanity of the repo at all times

- You need a folder tests/ with tests on your model and data loading for instance
- This should run from main root when you do `./scripts/smoke_test_all_papers.sh QTRL`

File: papers/QTRL/README.md [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+
+## ## Implementation Details
+
+* **Optimization**: The optimizer updates both the quantum circuit parameters (interferometer phases) and the classical mapping weights simultaneously.
+* **Discounted Returns**: Returns are computed recursively ($G_t = r_t + \gamma G_{t+1}$) and standardized using Z-score normalization to stabilize training.
+* **Gradient Clipping**: Applied to prevent exploding gradients, which are common in hybrid quantum-classical models.
+* **Model Factory**: All models are instantiated through the `create_hybrid_model` interface for better modularity.
+---
+
+> **Note:** This project is part of a research reproduction effort. Results may vary based on the photonic backend and noise models used.
```
Comment (CassNot) [inline papers/QTRL/README.md]: You are not using any noise model so this is a bit contradictory

File: papers/QTRL/lib/util.py (hunk 8 above)
Comment (leithkarrai) [reply papers/QTRL/lib/util.py:11]: adressed by adding `-e git+https://github.com/mit-han-lab/torchquantum.git#egg=torchquantum` in the requirements.txt file

Comment (leithkarrai) [reply papers/QTRL/lib/util.py:11]: Added tests folder, and raned smoketests on local to see if it is taken. Few tests have to be skiped though, because of gymnasium internal functioning (to run properly, minigrid backend has to be imported in the same time as gymnasium, but the library minigrid is not used, so ruff consider it as an unused library)

File: papers/QTRL/configs/defaults.json (hunk 7 above)
Comment (leithkarrai) [reply papers/QTRL/configs/defaults.json:1]: Adressed by adding --layer_dim to defaults.json. I thought it would be better for the user.

File: papers/QTRL/lib/util.py (hunk 6 above)
Comment (leithkarrai) [reply papers/QTRL/lib/util.py:12]: Addressed by copying DQNN structure.

**Review by CassNot — APPROVED** [review]
@leithkarrai , I added code to reproduce Figure 1 in the paper, run it and used the figure in your readme
I made the readme and notebook more self-explanatory and referenced the original work

**Comment by CassNot** [comment]
@leithkarrai, I had to rewrite the commit history to remove all the files from torchquantum (git rm does not erase history, so the history on this PR was quite heavy)

**Review by CassNot — APPROVED** [review]
